syntax = "proto3";

package nftp; // N1QL Functions Transport Protocol

message Config {
    uint32 WorkersPerNode = 1;
    uint32 ThreadsPerWorker = 2;
    string NsServerUrl = 3;
}

message Info {
    bool IsFatal = 1;
    string Message = 2;
}

message Void {
    uint32 Value = 1;
}

message Port {
    string EvaluatorId = 1;
    uint32 Port = 2;
}

message Function {
    string ID = 1;
    string Code = 2;
}

message Log {
    string EvaluatorID = 1;
    string Msg = 2;
}

message EvaluateRequest {
    string FunctionID = 1;
    string FunctionName = 2;
    string ThreadID = 3;
}

service Notification {
    rpc NotifyPort (Port) returns (Void);
    rpc Logger (Log) returns (Void);
}

service Evaluator {
    rpc Initialize (Config) returns (Info);
    rpc AddFunction (Function) returns (Info);
    rpc Evaluate (EvaluateRequest) returns (Info);
    rpc Stop (Void) returns (Void);
}

// Commands to run
// cd n1ql-functions
// protoc -I ./protobuf --grpc_out=../gen/nftp/server/ --plugin=protoc-gen-grpc=/Users/gautham/projects/github/grpc/bins/opt/grpc_cpp_plugin ./protobuf/evaluator.proto
// protoc -I ./protobuf --cpp_out=../gen/nftp/server/ ./protobuf/evaluator.proto
// protoc -I ./protobuf --go_out=plugins=grpc:../gen/nftp/client/ --plugin=protoc-gen-go=/Users/gautham/projects/github/grpc/third_party/protoc-gen-validate/vendor/github.com/golang/protobuf/protoc-gen-go/protoc-gen-go ./protobuf/evaluator.proto
